### ISSUE TITLE
Unevaluated function in CheckImplication condition

### ISSUE THREAD

**Issue by virgil-serbanuta** [kframework/kore#2793]
To reproduce:

tmp.k:
```
module TMP
  imports INT
  imports MAP

  syntax MyList ::= "nil"
                |   cons(Int, MyList)

  syntax KItem ::= iterate(MyList, Map)

  rule iterate(nil, _) => 0
  rule iterate(cons(Head:Int, Tail:MyList), M:Map)
        => splitMap(Head, M) ~> iterate(Tail, M)

  syntax KItem ::= splitMap(key:KItem, Map)

  rule splitMap(K:KItem, M:Map) => .K
    requires K in_keys(M)
    [label(splitMapInKeys)]
  rule splitMap(K:KItem, M:Map) => .K
    requires notBool K in_keys(M)
    [label(splitMapNotInKeys)]

  syntax Int ::= iterateFunction(MyList, Map)  [function, functional, smtlib(iterateFunction)]

  rule iterateFunction(nil, _:Map) => 0
  rule iterateFunction(cons(Head:Int, Tail:MyList), M:Map)
      => iterateFunction(Tail, M)
    requires notBool Head in_keys(M)
  rule iterateFunction(cons(Head:Int, Tail:MyList), M:Map)
      => iterateFunction(Tail, M)
    requires Head in_keys(M)

endmodule
```

tmp-proof.k:
```
module TMP-PROOF
  imports TMP

  claim iterate(L:MyList, M:Map) => iterateFunction(L, M)

endmodule
```

bash:
```
kompile tmp.k --backend haskell && kprove tmp-proof.k
```

Will print:
```
kore-exec: [465355] Warning (WarnStuckClaimState):
    (InfoReachability) while checking the implication:
    The proof has reached the final configuration, but the claimed implication is not valid. Location: /mnt/data/runtime-verification/elrond-multisig/tmp1/tmp-proof.k:4:9-4:58
  #Not ( {
    iterateFunction ( cons ( Head , Tail ) , M )
  #Equals
    iterateFunction ( Tail , M )
  } )
#And
  <k>
    iterateFunction ( Tail , M ) ~> _DotVar1 ~> .
  </k>
#And
  {
    true
  #Equals
    Head:Int in_keys ( M )
  }
```

Note that `iterateFunction ( cons ( Head , Tail ) , M )` should evaluate to `iterateFunction ( Tail , M )` when `Head:Int in_keys ( M )`. (in case you get the error on the other branch, `iterateFunction` should evaluate in the same way).

**Comment by JKTKops** [comment]
I'm at least 80% sure this is the same issue as #2785.

**Comment by ana-pantilie** [comment]
> I'm at least 80% sure this is the same issue as #2785.

That's possible, but, I think there was also the issue that the equation wasn't even attempted, right @virgil-serbanuta? If so, we should definitely see why that is.

**Comment by virgil-serbanuta** [comment]
That is correct.

FWIW, the backend is at commit `0521b0ea0ca6c2f61192822edf45a2ace7a52310`. Here are the `repl` commands I used for checking that the backend does not attempt to simplify the function:
```
step
select 2
step
select 3
step
log debug [DebugApplyEquation,DebugAttemptEquation] file eq.log disable-log-timestamps
step
```

Info for myself: The original proof is, I think, `protocol-correctness/proof/lemmas/list/contains/proof-last-distinct.k`

**Comment by virgil-serbanuta** [comment]
Info for myself: The original proof is actually `protocol-correctness/proof/functions/proof-count-can-sign.k`

**Comment by andreiburdusa** [comment]
> I'm at least 80% sure this is the same issue as #2785.

First time I tried running it on [this branch](https://github.com/kframework/kore/pull/2819) I used the wrong command and the desired backend was not actually used.
This time I gave the whole `kore-exec` path to `--haskell-backend-command` and the result was `#Top`. I got a bit confused and thought the whole path is not needed since the desired `kore-exec` was in my PATH.
```
andrei@q:~/Desktop/issue-2793$ kompile tmp.k --backend haskell && kprove tmp-proof.k --haskell-backend-command "/home/andrei/kore/.build/kore/bin/kore-exec"
[Warning] Compiler: Could not find main syntax module with name TMP-SYNTAX in
definition.  Use --syntax-module to specify one. Using TMP as default.
#Top
```
Sorry for the wasted time

**Comment by ana-pantilie** [comment]
This should be fixed on ed422630f9383b2774f1ec8ea04eee86ea0afe1e. Please let me know if you still encounter issues.
